proposals/0000-unsaturated-type-families.rst

```diff
@@ -309,17 +426,147 @@ the user needs to provide either a concrete matchability or a wildcard before
 supplying the instantiation for ``f``, as in ``qux @_ @Id``. This is tiresome,
 because ``m`` can *always* be inferred from the kind of ``f``, so it would be
 preferable to write ``qux @Id`` instead.
+
 The `explicit specificity <https://github.com/ghc-proposals/ghc-proposals/pull/99>`_
-proposal would make this possible.
+feature greatly improves the usability of unsaturated type families, as now the signature
+can be written as ::
+
+   qux :: forall {m} (f :: * -> @m *) a. f a -> f a
+
+Standalone Kind Signatures
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``StandaloneKindSignatures`` feature has the largest user-facing interaction
+surface with the current proposal. Before ``StandaloneKindSignatures``, the kind
+of a type family was never written in the source language, and was only hinted at by
+GHCi. Today this is no longer the case ::
+
+  type Id :: Type -> Type
+  type family Id a where
+    Id a = a
+
+The meaning of this kind signature is ``type Id :: Type -> @U Type``. However,
+requiring the user to spell out the full signature incurs a considerable amount
+of mental burden, so instead we propose to infer it from the definition.
```

Comment:
It's a very important property of SAKSs that we don't look at the definition! I think it should always use some sort of defaulting strategy, but it may depend on the tycon flavour (type family vs data family).

Reply:
Can you expand on why it's an important property? I think my proposal here is a modest extension (though I didn't want to fill the document with implementation details):
1) Take the SAKS as the user has written it, then generalise it (as it happens today).
2) Instantiate the arrows with matchability metavariables
3) Check the definition against the SAKS+matchability meta. In the process, we unify some variables and learn more information about them.
4) Zonk then generalise. This last step only quantifies over matchability variables as everything else was already fixed.

And then we get the full SAKS. Could this break something?

Reply:
The SAKS defines the final and complete kind of a type constructor, no metavariables. When you do step 3 of your plan, I assume you're going to have the type constructor added to the environment; and if you add a type constructor with a SAKS to the environment before you computed the final kind, you've pretty much defeated the entire purpose of a SAKS.

It will most likely interfere with mutually recursive definitions, such as:

```
type T1 :: ...
type T2 :: ...

data T1 a b c d = MkT1 (... use T2 here ...)
data T2 a b c d = MkT2 (... use T1 here ...)
```

The order of checking here is:

1. SAKS for T1
2. SAKS for T2
3. Definition of T1
4. Definition of T2

And it's quite important that when we're checking the definitions of T1 and T2, we already have the final kinds of these type constructors, as type/kind inference for mutually recursive definitions can't be possibly correct (there's some theoretical result about it, which I think somebody more knowledgable in type theory could cite).

Furthermore, type theory aside, checking SAKS without looking at the definition is also an important part of solving [`#12088`](https://gitlab.haskell.org/ghc/ghc/-/issues/12088), with the plan outlined here: https://gitlab.haskell.org/ghc/ghc/-/wikis/Type-&-Class-Dependency-Analysis

So, to summarize, this change would be:

1. against the core idea of SAKS
2. (most likely) break mutually recursive definitions
3. interfere with the current plan of addressing 12088

Reply:
Thanks for the pointers! I think it would be good to see some tricky examples of mutual recursion, as the ones I've tried all seemed to work fine. Of course the issue is that we can't infer principal types for mutually recursive groups, as that would be the same (I think) as inferring polymorphic recursion, thus undecidable. But maybe not inferring principal matchabilities is fine... I will try to think of some examples.

Another question is -- just when does matchability information affect the kind checking algorithm? I think when we have a non-generative constructor involved (i.e. a type family) that is also indexed in a matchability. But in that case, the proposed dependency analysis at https://gitlab.haskell.org/ghc/ghc/-/wikis/Type-&-Class-Dependency-Analysis already adds a dependency on the definition. Besides, my proposed algorithm never infers indexed matchabilities, so the type family in question would need to have a full signature including matchability information. Maybe this is fine? I'll need to think more about that kind checking algorithm to draw a conclusion.